network: include Chunk Validators in Tier1

saketh-are opened the pull request:
Adds chunk validators to the Tier1 participant set. In mainnet, T1 will grow in size from 100 to 250 participants.

Changing the function which calculates the participant set should be sufficient to implement this change. The code in the network crate which initiates outbound connections and accepts incoming ones operates according to the calculated participant set.

-----

OptimisticBlock was being broadcast over Tier1 to all peers. The CVs are not interested in receiving OB, so the behavior is changed to send only to BPs.**

The following message types sent over T1 are not affected in any way:
- `Tier1Handshake`, `HandshakeFailure`, and `LastEdge`, are used only to manage the connection.
- `BlockApproval` is routed specifically to the next block producer.
- `VersionedChunkEndorsement`  is routed specifically to the next few block producers.
- `VersionedPartialEncodedChunk` is routed to the block producers of current and next epochs.
- `ContractCodeRequest` is routed to one randomly chosen chunk producer.
- `ContractCodeResponse` is routed back to the author of the incoming request.

The following routed message types were already allowed on T1 and will now be sent over T1 more consistently:
- `PartialEncodedStateWitness` is routed in parts to all chunk validators.
- `PartialEncodedStateWitnessForward` is routed to all chunk validators.
- `ChunkContractAccesses` is routed to the chunk validators which aren't producers.


**I tested out sending the OB only to the next 5 block producers instead of sending it to all of them, but it caused `tests::optimistic_block::test_optimistic_block` to fail.

shreyan-gupta commented on chain/network/src/peer_manager/peer_manager_actor.rs:
```diff
@@ -783,12 +783,16 @@ impl PeerManagerActor {
+                // a conversion here from AccountId to peer connection. Consider reworking this.
+                let msg = Arc::new(PeerMessage::OptimisticBlock(optimistic_block));
+                for target_account in block_producers {
+                    if let Some(conn) = self.state.get_tier1_proxy_for_account_id(&target_account) {
```
Is it possible that we silently drop the message in case we are not able to find the tier1 proxy?

saketh-are replied:
Yes. Note that this is not a behavior change: OB was previously being broadcast to T1 peers only. 

If OB were a RoutedMessage instead of a PeerMessage we could use `send_message_to_account` which falls back to routing over T2 if the T1 connection is absent. But that change involves a more difficult migration and we will revisit this in the future.

shreyan-gupta commented on chain/client/src/client_actor.rs:
```diff
@@ -1381,8 +1383,25 @@ impl ClientActorInner {
+        let epoch_id = self
+            .client
+            .chain
+            .epoch_manager
```
nit: I guess we can add and additional line to get the epoch manager? `let epoch_manager = &self.client.chain.epoch_manager`?

shreyan-gupta commented on chain/network/src/peer_manager/network_state/tier1.rs:
```diff
@@ -374,4 +375,26 @@ impl super::NetworkState {
+    pub fn get_tier1_proxy_for_account_id(
+        &self,
+        account_id: &AccountId,
+    ) -> Option<Arc<connection::Connection>> {
```
We have almost exactly the same block in `send_message_to_account` function in chain/network/src/peer_manager/network_state/mod.rs

Might we want to merge/reuse this function there?

saketh-are replied:
They are very subtly different and I would prefer to limit blast radius of this PR. As mentioned in other comments we can revisit the exact handling of OptimisticBlock in the future and possibly convert it to a routed message. In that case we can delete this function.

wacban commented:
> send only to BPs.

@VanBarbascu please correct me if I'm wrong, I think the OB should be sent to CPs, not BPs? Or both to be safe? I do realize today they are the same but in the future we may want decouple those roles.

wacban commented on chain/client/src/client.rs:
```diff
@@ -2237,6 +2237,10 @@ impl Client {
         // block & chunk producers of the next epoch is too expensive, we can postpone it
         // till almost the end of this epoch.
+        //
+        // We include also all chunk validators of the current epoch, due to the importance of:
```
Can you also adjust this method's comment?

wacban commented on chain/client/src/client.rs:
```diff
@@ -2257,6 +2261,9 @@ impl Client {
                     .insert(bp.public_key().clone());
             }
         }
+        for v in self.epoch_manager.get_epoch_all_validators(&tip.epoch_id)? {
+            account_keys.entry(v.account_id().clone()).or_default().insert(v.public_key().clone());
+        }
```
Why not also do it for the next epoch?

saketh-are replied:
I guess I am wary of increasing the size of the T1 set more aggressively than we need to. A few missed endorsements at the start of the new epoch while new T1 connections are established should be a relatively minor issue. At the same time, if we think the CV set is generally quite stable and this doesn't really increase the connection count, I am fine with including next epoch CVs.

wacban replied:
TBH I would just make it consistent with the BPs and CPs to avoid any confusion around this in the future. Yes, I think we can safely assume this set is quite stable over time.

wacban commented on chain/client/src/client_actor.rs:
```diff
@@ -1381,8 +1383,25 @@ impl ClientActorInner {
+            .chain
+            .epoch_manager
+            .get_epoch_id_from_prev_block(optimistic_block.prev_block_hash())?;
+        //  TODO(#10584): Maybe we just need to send this to the next producers.
+        let block_producers: HashSet<AccountId> = self
```
yes, but chunk producers, not block ptoducers

VanBarbascu replied:
To address the limit case of epoch boundary, we want to send to the intersection of this and next epoch producers. 

This means that the node will always send to CV waiting to become producers, but that is fine as they will also benefit from OB because they are keeping up with the chain.

wacban commented on integration-tests/src/env/setup.rs:
```diff
@@ -618,16 +618,18 @@ fn process_peer_manager_message_default(
-                    .with_span_context(),
-                );
+            for (i, name) in validators.iter().enumerate() {
+                if block_producers.contains(name) {
+                    connectors[i].client_actor.do_send(
+                        OptimisticBlockMessage {
+                            optimistic_block: optimistic_block.clone(),
+                            from_peer: PeerInfo::random().id,
+                        }
+                        .with_span_context(),
+                    );
+                }
```
mini nit1: invert the condition to continue early and reduce nesting
mini nit2: move the message to a temporary variable to not clump too many things into a single line

both optional and based on personal preference ;)

wacban commented on test-loop-tests/src/tests/optimistic_block.rs:
```diff
@@ -253,7 +253,9 @@ fn alter_optimistic_block_at_height(
             let validator_signer = signer.clone();
             move |request: NetworkRequests| {
-                if let NetworkRequests::OptimisticBlock { optimistic_block } = &request {
+                if let NetworkRequests::OptimisticBlock { block_producers, optimistic_block } =
```
mini nit again about early returns:
```
let OB { .. } = &request else { return Some(request); };
if optimistic_block.height() != height { return Some(request); }
...
return ...
```

saketh-are commented:
@wacban @VanBarbascu I changed the Optimistic Block targets calculation to include chunk producers of the current and next epochs. I also added some caching behavior over it. PTAL.

shreyan-gupta reviewed (approved):
Approving for unblocking, but otherwise it would be a good idea to get a review from someone for OB.

wacban commented on chain/client/src/client.rs:
```diff
@@ -2257,11 +2264,39 @@ impl Client {
+        tip: &Tip,
+    ) -> Result<Arc<Vec<AccountId>>, Error> {
+        match &self.optimistic_block_targets_cache {
+            Some(it) if it.0 == tip.epoch_id => return Ok(it.1.clone()),
```
Can you get it by reference instead of cloning? 
nit: Can you unwrap it further?

wacban replied:
Ah I see it's an arc so clone should be cheap. Perhaps make it explicit with Arc::clone to make it less scary (guessing this syntax but there should be something like that).

wacban commented on chain/client/src/client.rs:
```diff
@@ -174,6 +174,8 @@ pub struct Client {
     upgrade_schedule: ProtocolUpgradeVotingSchedule,
     /// Produced optimistic block.
     last_optimistic_block_produced: Option<OptimisticBlock>,
+    /// Cached precomputed set of TIER1 accounts to which Optimistic Block should be sent.
+    optimistic_block_targets_cache: Option<(EpochId, Arc<Vec<AccountId>>)>,
```
Is it the targets just within this epoch or this and next? Either way can you mention that in the comment?

wacban replied:
I'd be tempted to try to remove the Arc and make it use references - though I'm not sure if that would be any better.

wacban commented on chain/client/src/client_actor.rs:
```diff
@@ -1381,8 +1381,13 @@ impl ClientActorInner {
         self.network_adapter.send(PeerManagerMessageRequest::NetworkRequests(
-            NetworkRequests::OptimisticBlock { optimistic_block: optimistic_block.clone() },
+            NetworkRequests::OptimisticBlock {
+                chunk_producers: (*targets).clone(),
+                optimistic_block: optimistic_block.clone(),
```
What do you think about wrapping those in `Arc`s to avoid the cloning?

saketh-are replied:
I made the change for `chunk_producers`. For `optimistic_block` further changes are needed on the network side to avoid the clone so I'm leaving it for now.

wacban commented on chain/network/src/peer_manager/network_state/tier1.rs:
```diff
@@ -374,4 +375,26 @@ impl super::NetworkState {
+    ) -> Option<Arc<connection::Connection>> {
+        let accounts_data = self.accounts_data.load();
+        for key in accounts_data.keys_by_id.get(account_id).iter().flat_map(|keys| keys.iter()) {
+            let data = match accounts_data.data.get(key) {
+                Some(data) => data,
+                None => continue,
+            };
```
nit: 
```suggestion
            let Some(data) = accounts_data.data.get(key) else { 
                continue;
            };
```

wacban commented on chain/network/src/peer_manager/network_state/tier1.rs:
```diff
@@ -374,4 +375,26 @@ impl super::NetworkState {
+                Some(data) => data,
+                None => continue,
+            };
+            let conn = match self.get_tier1_proxy(data) {
+                Some(conn) => conn,
+                None => continue,
+            };
```
ditto

wacban commented on chain/network/src/peer_manager/peer_manager_actor.rs:
```diff
@@ -783,12 +783,16 @@ impl PeerManagerActor {
-                    .tier1
-                    .broadcast_message(Arc::new(PeerMessage::OptimisticBlock(optimistic_block)));
+            NetworkRequests::OptimisticBlock { chunk_producers, optimistic_block } => {
+                // TODO(saketh): the block_producers are identified by their validator AccountId,
+                // but OptimisticBlock is sent over a direct PeerMessage. Hence we have to perform
+                // a conversion here from AccountId to peer connection. Consider reworking this.
```
nit: block_producers -> chunk_producers

wacban reviewed (approved):
LGTM

VanBarbascu commented on chain/client/src/client.rs:
```diff
@@ -2257,11 +2264,39 @@ impl Client {
+            _ => {}
+        }
+
+        let _guard =
```
```suggestion
        let _span =
```

VanBarbascu commented on chain/client/src/client.rs:
```diff
@@ -174,6 +174,8 @@ pub struct Client {
     /// Produced optimistic block.
     last_optimistic_block_produced: Option<OptimisticBlock>,
+    /// Cached precomputed set of TIER1 accounts to which Optimistic Block should be sent.
+    optimistic_block_targets_cache: Option<(EpochId, Arc<Vec<AccountId>>)>,
```
Instead of calling it optimistic block target, I would use something generic like: producers to make it clear where we are sending the OBs.
This is a subset of `tier1_accounts_cache`. Maybe we can combine them?

VanBarbascu commented on chain/client/src/client.rs:
```diff
@@ -2257,11 +2264,39 @@ impl Client {
                     .insert(bp.public_key().clone());
             }
         }
+        for v in self.epoch_manager.get_epoch_all_validators(&tip.epoch_id)? {
+            account_keys.entry(v.account_id().clone()).or_default().insert(v.public_key().clone());
+        }
```
Any chance we can add this logic only for Chunk Producers and Block Producers? 
This way CV will only connect to CP and BP and not to other CV.

saketh-are replied:
I'd prefer to keep it simple, considering that this optimization does not reduce the "worst case" number of connections.

saketh-are replied:
I think I'd prefer to keep it simple, considering that this optimization does not reduce the worst case number of connections.

VanBarbascu reviewed (commented):
A bit concerned about duplicating the logic of `tier1_accounts_cache` with `optimistic_block_targets_cache`.

We could combine them all under `tier1_accounts_cache` and have an object `T1AccountsCache{epoch_id: EpochId, bp: Arc<...>,cp: Arc<...>, cv: Arc<...>}` and functions for `getAll` and `getProducers`.

saketh-are commented:
Seems like upgradable pytest is still flaky...

wacban commented on chain/client/src/client.rs:
```diff
@@ -2266,12 +2269,43 @@ impl Client {
                     .or_default()
                     .insert(bp.public_key().clone());
             }
+            for v in self.epoch_manager.get_epoch_all_validators(&tip.epoch_id)? {
```
This should be `epoch_id`, not `tip.epoch_id`

wacban replied:
#13546 to fix it
